src/vptranslator.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2017 Kisio Digital and/or its affiliates.
+//
+// This program is free software: you can redistribute it and/or
+// modify it under the terms of the GNU General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+//use crate::common_format;
+
+use chrono::{Datelike, Duration, NaiveDate, Weekday};
+use num_traits::FromPrimitive;
+use std::collections::BTreeSet;
+use std::vec::Vec;
+
+#[derive(Debug, PartialOrd)]
+struct ValidityPeriod {
+    start_date: NaiveDate,
+    end_date: NaiveDate,
+}
+
+impl Default for ValidityPeriod {
+    fn default() -> Self {
+        ValidityPeriod {
+            start_date: chrono::naive::MIN_DATE,
+            end_date: chrono::naive::MIN_DATE,
+        }
+    }
+}
+
+impl PartialEq for ValidityPeriod {
+    fn eq(&self, other: &Self) -> bool {
+        (self.start_date, &self.end_date) == (other.start_date, &other.end_date)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum ExceptionType {
+    Sub,
+    Add,
+}
+
+#[derive(Debug)]
+struct ExceptionDate {
+    date: NaiveDate,
+    exception_type: ExceptionType,
+}
+
+impl PartialEq for ExceptionDate {
+    fn eq(&self, other: &ExceptionDate) -> bool {
+        self.exception_type == other.exception_type
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct BlockPattern {
+    week: u8,
+    nb_weeks: i64,
+    operating_days: Vec<Weekday>,
+    validity_periods: Vec<ValidityPeriod>,
+    exceptions: Vec<ExceptionDate>,
+}
+
+fn get_prev_monday(date: NaiveDate) -> NaiveDate {
+    let res = date + Duration::days(-1 * date.weekday().num_days_from_monday() as i64);
+    res
+}
+
+fn weeks(dates: &BTreeSet<NaiveDate>) -> Vec<u8> {
+    let start_date: NaiveDate = get_prev_monday(*dates.iter().next().unwrap());
+    let end_date: NaiveDate = *dates.iter().next_back().unwrap();
+    let length = (end_date.signed_duration_since(start_date).num_weeks() + 1) as usize;
+
+    let mut res = vec![0; length];
+    for date in dates {
+        let w = date.signed_duration_since(start_date).num_weeks() as usize;
+        res[w] |= 1 << (7 - date.weekday().number_from_monday());
+    }
+    res
+}
+
+fn dists(w: u8, weeks: &Vec<u8>) -> u32 {
+    let mut res: u32 = 0;
```

Comment:
it's completly minor and the way you wrote it is perfect, but just so you know I think you could have written it like

```rust
    weeks.iter()
    .filter(|n| **n != 0u8)
    .map(|n| (*n as u8 ^ w).count_ones())
    .sum()
```